library/pkcs7.c

```diff
@@ -57,7 +57,10 @@ static int pkcs7_get_next_content_len(unsigned char **p, unsigned char *end,
     ret = mbedtls_asn1_get_tag(p, end, len, MBEDTLS_ASN1_CONSTRUCTED
                                | MBEDTLS_ASN1_CONTEXT_SPECIFIC);
     if (ret != 0) {
-        ret = MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_FORMAT, ret);
+        ret = MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO, ret);
+    } else if ((size_t) (end - *p) != *len) {
```

Comment:
Should we add the `else `branch to make it safer or program is as what the line 192 and 195 do?

Reply:
I am not totally clear what you are suggesting (re-structuring the if statements - but how exactly?), but I don't see a concern here. As it stands, the conditions are fairly simple and easy to follow IMO.

Reply:
> I am not totally clear what you are suggesting (re-structuring the if statements - but how exactly?), but I don't see a concern here. As it stands, the conditions are fairly simple and easy to follow IMO.

I just remember that the requirement for a final else statement is defensive programming.

The else statement should either take appropriate action or contain a suitable comment as to why no action is taken. This is consistent with the requirement to have a final default clause in a switch statement.  See https://rules.sonarsource.com/c/RSPEC-126 as one example. 

Actually, I can't quite remember the attack method base on the case "else if without else"(I will try to figure out the way, but maybe can't get the result recently), but I think maybe it's a better way if we apply the defensive programming.

The suggested way is:
```
if (ret != 0) {
}
if ((size_t) (end - *p) != *len) {
}
```

Reply:
Err, no, we definitely don't want to do that. The code is right - we want to set our `ret` to one value if `mbedtls_asn1_get_tag()` fails, and to another if we aren't exactly at the end. "The suggested way" changes those semantics.

Another way of looking at it is

```
if (ret == 0) {
    if ((size_t) (end - *p) != *len) {
        ret = MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO,
            MBEDTLS_ERR_ASN1_LENGTH_MISMATCH);
    }
} else {
    ret = MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO, ret);
}
```

Reply:
> Err, no, we definitely don't want to do that. The code is right - we want to set our `ret` to one value if `mbedtls_asn1_get_tag()` fails, and to another if we aren't exactly at the end. "The suggested way" changes those semantics.

Aha, you are right. We didn't have `goto `or `return `in the **if...else** condtion.

Reply:
That standard suggests we could do this by adding a comment like:

```
} else {
    /* Everything looks OK - we will return the value of ret returned by mbedtls_asn1_get_tag() */
}
```

but we don't generally aim to follow MISRA standards in the code-base. Discussed and all agreed no change needed.